lib/localisation/src/preprocessing/accelerometer.rs

```diff
@@ -1 +1,300 @@
+    q1: f32,
+    q2: f32,
+    q3: f32,
+
+    iqr: f32,
+    lower_bound: f32,
+    upper_bound: f32,
+}
+
+impl Quartiles {
+    pub fn new(q1: f32, q2: f32, q3: f32, is_unreliable: bool) -> Self {
+        let mut bound_factor = 1.5;
+        if is_unreliable {
+            bound_factor = 1.2;
+        }
+
+        Self {
+            q1,
+            q2,
+            q3,
+            iqr: q3 - q1,
+            lower_bound: q1 - bound_factor * (q3 - q1),
+            upper_bound: q3 + bound_factor * (q3 - q1),
+        }
+    }
+}
+
+pub struct AccelerometerPreprocessor {
+    num_reliable_accelerometers: i32,
+    reliable_accelerometers: [bool; K_NUM_ACCELEROMETERS],
+    num_outliers_per_accelerometer: [i32; K_NUM_ACCELEROMETERS],
+}
+
+impl Default for AccelerometerPreprocessor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AccelerometerPreprocessor {
+    pub fn new() -> Self {
+        Self {
+            num_reliable_accelerometers: K_NUM_ACCELEROMETERS as i32,
+            reliable_accelerometers: [true; K_NUM_ACCELEROMETERS],
+            num_outliers_per_accelerometer: [0; K_NUM_ACCELEROMETERS],
+        }
+    }
+
+    pub fn handle_outliers(
+        &mut self,
+        data: AccelerometerData<K_NUM_ACCELEROMETERS>,
+    ) -> Option<AccelerometerData<K_NUM_ACCELEROMETERS>> {
+        let quartiles = match self.calculate_quartiles(&data) {
+            Some(quartiles) => quartiles,
+            None => return None,
+        };
+
+        let accelerometer_data = data
+            .iter()
+            .enumerate()
+            .map(|(i, &val)| {
+                if !self.reliable_accelerometers[i] {
+                    quartiles.q2
+                } else if val < quartiles.lower_bound || val > quartiles.upper_bound {
+                    self.num_outliers_per_accelerometer[i] += 1;
+                    quartiles.q2
+                } else {
+                    self.num_outliers_per_accelerometer[i] = 0;
+                    val
+                }
+            })
+            .collect();
+
+        Some(accelerometer_data)
+    }
+
+    pub fn calculate_quartiles(
+        &self,
+        data: &AccelerometerData<K_NUM_ACCELEROMETERS>,
+    ) -> Option<Quartiles> {
+        let quartiles: Quartiles;
+        if self.num_reliable_accelerometers == K_NUM_ACCELEROMETERS as i32 {
+            quartiles = self.get_quartiles(data);
+        } else if self.num_reliable_accelerometers == (K_NUM_ACCELEROMETERS as i32 - 1) {
+            const SIZE: usize = K_NUM_ACCELEROMETERS - 1;
+            let mut filtered_data: AccelerometerData<SIZE> =
+                AccelerometerData::from_slice(&[0.0; SIZE]).unwrap();
+            let mut filtered_data_idx = 0;
+            data.iter().enumerate().for_each(|(i, val)| {
+                if self.reliable_accelerometers[i] {
+                    filtered_data[filtered_data_idx] = *val;
+                    filtered_data_idx += 1;
+                }
+            });
```

Comment:
```rust
let filtered_data: AccelerometerData<SIZE> = data
    .iter()
    .enumerate()
    .filter(|(i, _)| self.reliable_accelerometers[*i])
    .map(|(_, &val)| val)
    .collect();
```
this might be nicer (if i understand correctly what you're doing) (and if i remember how rust works)